### ISSUE TITLE
Potential whisper.cpp GPU support via the Const-me Windows Implementation

### ISSUE THREAD

**Issue by albino1** [SubtitleEdit/subtitleedit#6651]
I'm sure you're already following this project, but in case you're not, this guy has created a Windows GUI implementation of whisper.cpp with GPU support:

https://github.com/Const-me/Whisper

It doesn't currently support video files, only audio files, but I tested it by extracting a WAV file from a 93 minute MKV and then running it through his GUI with `medium.en.bin` and it was _exceptionally_ faster than even the OpenBLAS method:

<details>
  <summary>Output Screenshot</summary>

![93](https://user-images.githubusercontent.com/12164957/216792715-dc740f6f-6c66-4170-9265-5c331797c74f.jpg)

</details>

It completed in only 5:39 as compared to what would probably be hours using the normal method.

As far as accuracy goes... well I don't know, I haven't really tested it that thoroughly, but certainly it's worth investigating.

Anyway, just wanted to let you know in case you were unaware.  Hopefully this is something Subtitle Edit could take advantage of in the future.  Thanks!

**Comment by darnn** [comment]
Works great here, both in terms of speed and in terms of accuracy (at least with the one file I tried, which was in English, and using large.bin).

**Comment by rsmith02ct** [comment]
I was about to ask the same thing- I'm finding the CPU only implementation quite slow here as each of my interviews is at least an hour long. The overall SubtitleEdit interface is great so combining that with GPU support would be the best of both worlds.

**Comment by rsmith02ct** [comment]
Also it does support video files- at least AVC in an MP4 wrapper,  I just processed a few at about 30 fps on a NVIDIA 1050.

**Comment by niksedk** [comment]
I does not give any performance improvements here... with my intel gpu ,;)
But it should be relatively simple to add "whisper cpp gpu" to SE.

**Comment by rsmith02ct** [comment]
That would be a massive change. Happy to do a performance comparison of it with my two systems (NVIDIA 1050 and 2080) but it looks like at least a 6x speed increase.

**Comment by albino1** [comment]
In fairness, I'm using an RTX 4080, which is probably overkill for this task, but it's _really_ fast for me by comparison.  I look forward to checking it out, thanks @niksedk!

**Comment by niksedk** [comment]
Tried to make a test version here: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.11/SubtitleEditBeta.zip
I was not able to use anything but tiny/small model... might be due to my intel gpu mem...

Does it work for any of you? (you must switch to "ConstMe")
![image](https://user-images.githubusercontent.com/1261331/217328867-cf398a1f-bfba-4ccc-9acd-69bcc8b7d466.png)

**Comment by darnn** [comment]
Tried it with large.bin and a 40-second Hebrew file. The only output was a single period, no text. But at least it was quick!

**Comment by niksedk** [comment]
> Tried it with large.bin and a 40-second Hebrew file. The only output was a single period, no text. Same with tiny. But at least it was quick!

Did you check the error_log.txt ?

**Comment by albino1** [comment]
That was quick!

Worked great for me with English.

Testing it on the same 93 minute test file as last time, but now using the original Blu-ray remux MKV, and not just an extracted WAV file, I get:

`tiny.en`: 1:32 with 79% GPU usage and 2.2gb GPU memory

`small.en`: 2:26 with 88% GPU usage and 2.6gb GPU memory

`medium.en`: 5:59 with 92% GPU usage and 3.9gb GPU memory

`large`: 7:19 with 96% GPU usage and 5.7gb of GPU memory

I'd test it with regular whisper.cpp, but it would be pretty annoying to sit here with a timer running for potentially hours waiting for the finished okay button :)

Side note: I'm not sure if this makes sense with Subtitle Edit, or would shorten the run time, but this pull request on whisper.cpp to suppress non-speech stuff would be nice to have as well once it gets merged:

https://github.com/ggerganov/whisper.cpp/pull/473

For example, I have like 10+ straight entries for `[birds chirping]` and so many versions of just `[music]` all over the place, hah.  I know we can clean it up in Subtitle Edit, but it would be nice to just exclude it from the get go via a toggle.

**Comment by darnn** [comment]
> Did you check the error_log.txt ?

It didn't create one.

**Comment by rsmith02ct** [comment]
This is a huge improvement. Even on my 4GB NVIIDA 1050 (mobile) GPU I am getting much faster results with base, small and even medium models.  Only tested on a 3 min video so far and will test on other videos and with my desktop later.

**Comment by niksedk** [comment]
Nice it works somewhat :)

@darnn: Is latest beta writing anything to error_log.txt ?
https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.11/SubtitleEditBeta.zip
(you can open the "SE Data folder" where error_log.txt should be with ctrl+shift+alt+D in the SE main window - but you probably know that)
SE just download the binaries from https://github.com/Const-me/Whisper  (version 1.7)

**Comment by darnn** [comment]
Seems to work with English now. With Hebrew, it appears to recognize the audio, but the output is all question marks. Here's the relevant portion of error_log.txt:
[error log excerpt2.txt](https://github.com/SubtitleEdit/subtitleedit/files/10683356/error.log.excerpt2.txt)
And here's the audio:
[00_00_00_00__00_01_00_00.zip](https://github.com/SubtitleEdit/subtitleedit/files/10683361/00_00_00_00__00_01_00_00.zip)

**Comment by MbuguaDavid** [comment]
> Tried to make a test version here: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.11/SubtitleEditBeta.zip I was not able to use anything but tiny/small model... might be due to my intel gpu mem...
> 
> Does it work for any of you? (you must switch to "ConstMe") ![image](https://user-images.githubusercontent.com/1261331/217328867-cf398a1f-bfba-4ccc-9acd-69bcc8b7d466.png)

It works on my end and it's 4x faster than Whisper CPP. WOW!!! 
https://youtu.be/McN6PtdLWYI

**Comment by niksedk** [comment]
@darnn: thx for the files - hopefully fixed here: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.11/SubtitleEditBeta.zip

**Comment by darnn** [comment]
Seems to be! Tested with Hebrew and Russian. There are only two minor things:
The results seem to be slightly different than those I get in WhisperDesktop itself. You can see it in the file I attached above, but here's a clearer example, one sentence from it (that both handle poorly because it's hard to parse in general):
[a14cd8df-068a-466f-976a-702d75fb2a90.zip](https://github.com/SubtitleEdit/subtitleedit/files/10689667/a14cd8df-068a-466f-976a-702d75fb2a90.zip)
And the way the text is broken up into individual subtitles is also different (can be seen in the minute-long file as well).

The other thing is that in SE's output, there will sometimes not be a space after a period. The first line from the minute-long file above:
לפני כמה שנים חברת וולבו השוודית הגדולה והעוצמתית נקלעה לקשיים.הפתרון היה שחברת ג'ילי הסינית רכשה אותה.

**Comment by rsmith02ct** [comment]
![Screenshot 2023-02-09 11 30 17](https://user-images.githubusercontent.com/122251109/217702519-6b36d6cf-809a-4e8e-941e-2b6a7eab8e75.png)

Also tested on a Japanese video and the captions mostly worked. It kept repeating one line for some reason and lost the timing at that point. (screenshot)

I tried it again with CPP and it didn't have that issue though timings and line breaks were mostly off.
Tried it a second time with Const and had the same repeated lines issue. 

![Screenshot 2023-02-09 11 49 39](https://user-images.githubusercontent.com/122251109/217705201-32735b0a-f704-4efb-bce9-aad2ff331847.png)

**Comment by albino1** [comment]
> It kept repeating one line for some reason and lost the timing at that point. (screenshot)

I have this problem with all non-English languages, but I also have that problem with the original Const-me using a WAV file, and with the original `whisper.cpp`, so I don't think it's Subtitle Edit related.  English seems to work fine, but other languages get stuck on some line.

For example, the Japanese one I mentioned earlier has 8 straight minutes of:

`- 時間は大丈夫? - 時間は大丈夫?`

And a Spanish one I tried has `¿Qué pasa?` repeated like a thousand times.

This is using `ggml-medium.bin` and `tiny.bin` from [Hugging Face](https://huggingface.co/datasets/ggerganov/whisper.cpp/tree/main) as recommended in the `whisper.cpp` repository, so maybe there's a different training set that works better for non-English languages that I'm unaware of.

**Comment by rsmith02ct** [comment]
Thanks for that report that I'm not the only one. Is it an issue inherent
to Whisper or specific to these implementations? I'm just wondering how far
before SubtitleEdit we need to go to try to get it resolved.

**Comment by darnn** [comment]
Oh yeah, if it was possible to affect those parameters somehow (especially in WhisperDesktop), I would love that, because that does also happen to me with it (and with whisper.cpp in general). It seems to happen less frequently with this, for whatever reason, though: https://colab.research.google.com/github/ANonEntity/WhisperWithVAD/blob/main/WhisperWithVAD.ipynb

**Comment by niksedk** [comment]
I think we get less repetitions with the large model. 

Hopefully this will be fixed in a future release (as well as more precise timestamps)

**Comment by albino1** [comment]
Thanks for the idea niksedk.  I gave the large a try, but still ended up with `¿Qué pasa?` repeated over and over for the last hour and 20 minutes of the movie, hah.

**Comment by niksedk** [comment]
OK, you could also chunk up the file... SE can also do Whisper on selected lines.

... and hoping for OpenAI to fix this :)

**Comment by eddilara** [comment]
can we buy you a nvidia and amd gpu so you can improve the gpu processing?

**Comment by darnn** [comment]
Not that I wouldn't pitch in, but like he said, it depends more on OpenAI (and Const-me, and any other GPU-enabled versions, like this one, which I haven't tried https://github.com/guillaumekln/faster-whisper ). SE can't make Whisper itself any faster than it is.
